OrbitLinuxTracing/PerfEventProcessorTest.cpp

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <thread>
+
+#include "LinuxTracingUtils.h"
+#include "PerfEventProcessor.h"
+#include "PerfEventVisitor.h"
+
+namespace LinuxTracing {
+
+namespace {
+
+class MockVisitor : public PerfEventVisitor {
+ public:
+  MOCK_METHOD(void, visit, (LostPerfEvent * event), (override));
+};
+
+class PerfEventProcessorTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    processor_.AddVisitor(&mock_visitor_);
+    processor_.SetDiscardedOutOfOrderCounter(&discarded_out_of_order_counter_);
+  }
+
+  void TearDown() override {}
+
+  PerfEventProcessor processor_;
+  MockVisitor mock_visitor_;
+  std::atomic<uint64_t> discarded_out_of_order_counter_ = 0;
+
+  static constexpr uint64_t kDelayBeforeProcessOldEventsMs = 100;
+};
+
+std::unique_ptr<PerfEvent> MakeFakePerfEvent(int origin_fd, uint64_t timestamp_ns) {
+  // We use LostPerfEvent just because it's a simple one, but we could use any
+  // as we only need to set the file descriptor and the timestamp.
+  auto event = std::make_unique<LostPerfEvent>();
+  event->SetOriginFileDescriptor(origin_fd);
+  event->ring_buffer_record.sample_id.time = timestamp_ns;
+  return event;
+}
+
+}  // namespace
+
+TEST_F(PerfEventProcessorTest, ProcessOldEvents) {
+  EXPECT_CALL(mock_visitor_, visit).Times(0);
+  processor_.AddEvent(MakeFakePerfEvent(11, MonotonicTimestampNs()));
+  processor_.AddEvent(MakeFakePerfEvent(11, MonotonicTimestampNs()));
+  processor_.AddEvent(MakeFakePerfEvent(22, MonotonicTimestampNs()));
+  processor_.ProcessOldEvents();
+
+  ::testing::Mock::VerifyAndClearExpectations(&mock_visitor_);
```

Comment:
Uh that is a nice feature, I wasn't aware of :)

Reply:
I used it extensively in the IPC tests. In time-sensitive code, it's the way I found to say "`EXPECT_CALL` by now".